src/graph/graph_op.cc

```diff
@@ -38,199 +42,208 @@ class RangeIter : public std::iterator<std::input_iterator_tag, dgl_id_t> {
-  IdArray indices_arr = aten::NewIdArray(num_edges);
-  IdArray edge_ids_arr = aten::NewIdArray(num_edges);
-  dgl_id_t* indptr = static_cast<dgl_id_t*>(indptr_arr->data);
-  dgl_id_t* indices = static_cast<dgl_id_t*>(indices_arr->data);
-  dgl_id_t* edge_ids = static_cast<dgl_id_t*>(edge_ids_arr->data);
-
-  indptr[0] = 0;
-  dgl_id_t cum_num_nodes = 0;
-  dgl_id_t cum_num_edges = 0;
-  for (const ImmutableGraph *gr : graphs) {
-    const CSRPtr g_csrptr = gr->GetInCSR();
-    const int64_t g_num_nodes = g_csrptr->NumVertices();
-    const int64_t g_num_edges = g_csrptr->NumEdges();
-    dgl_id_t* g_indptr = static_cast<dgl_id_t*>(g_csrptr->indptr()->data);
-    dgl_id_t* g_indices = static_cast<dgl_id_t*>(g_csrptr->indices()->data);
-    dgl_id_t* g_edge_ids = static_cast<dgl_id_t*>(g_csrptr->edge_ids()->data);
-    for (dgl_id_t i = 1; i < g_num_nodes + 1; ++i) {
-      indptr[cum_num_nodes + i] = g_indptr[i] + cum_num_edges;
+GraphPtr GraphOp::DisjointUnion(std::vector<GraphPtr> graphs) {
+  CHECK_GT(graphs.size(), 0) << "Input graph list is empty";
+  if (IsMutable(graphs[0])) {
+    // Disjointly union of a list of mutable graph inputs. The result is
+    // also a mutable graph.
+    MGraphPtr rst = Graph::Create();
+    uint64_t cumsum = 0;
+    for (GraphPtr gr : graphs) {
+      MGraphPtr mg = std::dynamic_pointer_cast<Graph>(gr);
+      CHECK(mg) << "All the input graphs should be mutable graphs.";
+      rst->AddVertices(gr->NumVertices());
+      for (uint64_t i = 0; i < gr->NumEdges(); ++i) {
+        // TODO(minjie): quite ugly to expose internal members
+        rst->AddEdge(mg->all_edges_src_[i] + cumsum, mg->all_edges_dst_[i] + cumsum);
+      }
+      cumsum += gr->NumVertices();
     }
-    for (dgl_id_t i = 0; i < g_num_edges; ++i) {
-      indices[cum_num_edges + i] = g_indices[i] + cum_num_nodes;
+    return rst;
+  } else {
+    // Disjointly union of a list of immutable graph inputs. The result is
+    // also an immutable graph.
+    int64_t num_nodes = 0;
+    int64_t num_edges = 0;
+    for (auto gr : graphs) {
+      num_nodes += gr->NumVertices();
+      num_edges += gr->NumEdges();
     }
+    IdArray indptr_arr = aten::NewIdArray(num_nodes + 1);
+    IdArray indices_arr = aten::NewIdArray(num_edges);
+    IdArray edge_ids_arr = aten::NewIdArray(num_edges);
+    dgl_id_t* indptr = static_cast<dgl_id_t*>(indptr_arr->data);
+    dgl_id_t* indices = static_cast<dgl_id_t*>(indices_arr->data);
+    dgl_id_t* edge_ids = static_cast<dgl_id_t*>(edge_ids_arr->data);
 
-    for (dgl_id_t i = 0; i < g_num_edges; ++i) {
-      edge_ids[cum_num_edges + i] = g_edge_ids[i] + cum_num_edges;
+    indptr[0] = 0;
+    dgl_id_t cum_num_nodes = 0;
+    dgl_id_t cum_num_edges = 0;
+    for (auto g : graphs) {
+      ImGraphPtr gr = std::dynamic_pointer_cast<ImmutableGraph>(g);
+      CHECK(gr) << "All the input graphs should be immutable graphs.";
+      // TODO(minjie): why in csr?
```

Comment:
Just for simplicity, because the construction of immutable graph uses incsr.

Reply:
I left a note here because the immutable graph might have only out_csr. In this case, the function will first convert it to in_csr, which is somewhat redundant because we could directly batch out_csr and construct the result graph using out_csr. However, I don't know what is a better solution here because this means the function needs to know the internal state of an immutable graph object, breaking the encapsulation.